File: examples/node/gas-price-monitor.js
/**
 * Gas Price Monitor Example
 * 
 * This example demonstrates real-time gas price monitoring
 * across multiple chains with fast/average/slow recommendations.
 * 
 * Use case: Transaction optimization, cost estimation
 */

const { CoinRailzClient } = require('../../sdk/typescript/coinrailz-client');

async function main() {
  const client = new CoinRailzClient({
    apiKey: process.env.COINRAILZ_API_KEY || 'cr_live_YOUR_API_KEY_HERE'
  });

  console.log('⛽ Real-time Gas Price Monitor\n');
  console.log('═'.repeat(70));

  // Monitor these chains
  const chains = ['ethereum', 'base', 'polygon', 'arbitrum', 'bnb'];

  try {
    const gasData = await client.getGasPrices(chains);

    if (gasData.success) {
      for (const [chain, prices] of Object.entries(gasData.data)) {
        console.log(`\n${chain.toUpperCase()}`);
        console.log('─'.repeat(70));
        console.log(`Fast (< 30s):     ${prices.fast} Gwei    (~$${estimateCost(prices.fast, chain)})`);
        console.log(`Average (< 2m):   ${prices.average} Gwei    (~$${estimateCost(prices.average, chain)})`);
        console.log(`Slow (< 10m):     ${prices.slow} Gwei    (~$${estimateCost(prices.slow, chain)})`);
        
        // Recommendation
        const recommendation = getRecommendation(prices);
        console.log(`\n💡 Recommendation: ${recommendation}`);
      }
    }

    console.log('\n' + '═'.repeat(70));

    // Balance check
    const balance = await client.getBalance();
    console.log(`\n💳 Credits: $${balance.balance.toFixed(2)}`);

  } catch (error) {
    console.error('❌ Error:', error.message);
  }
}

function estimateCost(gwei, chain) {
  // Estimate cost for standard ERC20 transfer (~65,000 gas)
  const gasUnits = 65000;
  const ethPrice = 2850; // Approximate
  const bnbPrice = 320;
  const maticPrice = 0.85;

  let cost;
  if (chain === 'ethereum' || chain === 'base' || chain === 'arbitrum') {
    cost = (gwei * gasUnits * ethPrice) / 1e9;
  } else if (chain === 'bnb') {
    cost = (gwei * gasUnits * bnbPrice) / 1e9;
  } else if (chain === 'polygon') {
    cost = (gwei * gasUnits * maticPrice) / 1e9;
  }

  return cost.toFixed(2);
}

function getRecommendation(prices) {
  const avgGwei = prices.average;
  
  if (avgGwei < 10) return '✅ Great time to transact - low fees';
  if (avgGwei < 30) return '👍 Good time to transact - moderate fees';
  if (avgGwei < 50) return '⚠️ Fees are elevated - consider waiting';
  return '🚨 Very high fees - wait if not urgent';
}

// Run the example
main().catch(console.error);

File: examples/python/gas_price_monitor.py
"""
Gas Price Monitor Example

This example demonstrates real-time gas price monitoring
across multiple chains with fast/average/slow recommendations.

Use case: Transaction optimization, cost estimation
"""

import os
import sys
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../sdk/python'))

from coinrailz_client import CoinRailzClient


def estimate_cost(gwei: float, chain: str) -> str:
    """Estimate transaction cost in USD"""
    # Estimate cost for standard ERC20 transfer (~65,000 gas)
    gas_units = 65000
    eth_price = 2850  # Approximate
    bnb_price = 320
    matic_price = 0.85

    if chain in ['ethereum', 'base', 'arbitrum']:
        cost = (gwei * gas_units * eth_price) / 1e9
    elif chain == 'bnb':
        cost = (gwei * gas_units * bnb_price) / 1e9
    elif chain == 'polygon':
        cost = (gwei * gas_units * matic_price) / 1e9
    else:
        cost = 0

    return f"{cost:.2f}"


def get_recommendation(prices: dict) -> str:
    """Get transaction recommendation based on gas prices"""
    avg_gwei = prices['average']
    
    if avg_gwei < 10:
        return '✅ Great time to transact - low fees'
    elif avg_gwei < 30:
        return '👍 Good time to transact - moderate fees'
    elif avg_gwei < 50:
        return '⚠️ Fees are elevated - consider waiting'
    else:
        return '🚨 Very high fees - wait if not urgent'


def main():
    client = CoinRailzClient(
        api_key=os.getenv('COINRAILZ_API_KEY', 'cr_live_YOUR_API_KEY_HERE')
    )

    print('⛽ Real-time Gas Price Monitor\n')
    print('═' * 70)

    # Monitor these chains
    chains = ['ethereum', 'base', 'polygon', 'arbitrum', 'bnb']

    try:
        gas_data = client.get_gas_prices(chains)

        if gas_data.get('success'):
            for chain, prices in gas_data['data'].items():
                print(f"\n{chain.upper()}")
                print('─' * 70)
                print(f"Fast (< 30s):     {prices['fast']} Gwei    (~${estimate_cost(prices['fast'], chain)})")
                print(f"Average (< 2m):   {prices['average']} Gwei    (~${estimate_cost(prices['average'], chain)})")
                print(f"Slow (< 10m):     {prices['slow']} Gwei    (~${estimate_cost(prices['slow'], chain)})")
                
                # Recommendation
                recommendation = get_recommendation(prices)
                print(f"\n💡 Recommendation: {recommendation}")

        print('\n' + '═' * 70)

        # Balance check
        balance = client.get_balance()
        print(f'\n💳 Credits: ${balance.balance:.2f}')

    except Exception as e:
        print(f'❌ Error: {str(e)}')


if __name__ == '__main__':
    main()
